core/src/test/java/org/apache/iceberg/TestPartitionSpecUpdate.java

```diff
@@ -0,0 +1,255 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TestPartitionSpecUpdate extends TableTestBase {
+
+  private int[] expectedFieldIds;
+
+  private String[] expectedSpecs;
+
+  @Parameterized.Parameters
+  public static Object[][] parameters() {
+    return new Object[][] {
+        new Object[] { 1, new int[]{ 1000, 1001, 1001, 1000, 1000 },
+            new String[]{"  1000: data_bucket_removed: void(2)\n", "  1002: id_bucket_removed: void(1)\n"} },
+        new Object[] { 2, new int[]{ 1001, 1000, 1001, 1002, 1002 }, new String[]{"", ""} },
+    };
+  }
+
+  public TestPartitionSpecUpdate(int formatVersion, int[] expectedFieldIds, String[] expectedSpecs) {
+    super(formatVersion);
+    this.expectedFieldIds = expectedFieldIds;
+    this.expectedSpecs = expectedSpecs;
+  }
+
+  @Test
+  public void testCommitUpdatedSpec() {
+    Assert.assertEquals("[\n" +
+        "  1000: data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(1000, table.spec().lastAssignedFieldId());
+
+    table.updateSpec().clear()
+        .addBucketField("id", 8)
+        .addBucketField("data", 16)
+        .commit();
+
+    Assert.assertEquals("[\n  " +
+        expectedFieldIds[0] + ": id_bucket: bucket[8](1)\n  " +
+        expectedFieldIds[1] + ": data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(expectedFieldIds[2], table.spec().lastAssignedFieldId());
+
+    table.updateSpec().clear()
+        .addTruncateField("data", 8)
+        .commit();
+
+    Assert.assertEquals("[\n  " +
+        expectedFieldIds[3] + ": data_trunc: truncate[8](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(expectedFieldIds[4], table.spec().lastAssignedFieldId());
+  }
+
+  @Test
+  public void testUpdateException() {
+    AssertHelpers.assertThrows(
+        "Should throw IllegalArgumentException if there is an invalid partition field",
+        IllegalArgumentException.class, "Cannot use partition name more than once: id_bucket",
+        () -> table.updateSpec().clear()
+            .addBucketField("id", 8)
+            .addBucketField("id", 16)
+            .commit());
+  }
+
+  @Test
+  public void testAddDuplicateFieldException() {
+    AssertHelpers.assertThrows(
+        "Should throw IllegalArgumentException if adding a duplicate partition field",
+        IllegalArgumentException.class, "Cannot use partition name more than once: data_bucket",
+        () -> table.updateSpec()
+            .addBucketField("data", 16)
```

Comment:
The duplicate partition detection should take precedence over duplicate name detection. In other words, I would expect this case to fail because there is already a bucket field for data, not because of the name duplication.

Reply:
👌  updated accordingly.